File: CrackingTheCodingInterview/4_trees_and_graphs/1_is_balanced.cpp
#include <algorithm>

struct Node
{
    Node* left;
    Node* right;
};

int getHeight(Node* node, bool& balanced)
{
    if (!node)
    {
        balanced = true;
        return 0;
    }

    bool balancedLeft;
    int heightLeft = getHeight(node->left, balancedLeft);
    bool balancedRight;
    int heightRight = getHeight(node->right, balancedRight);

    balanced = balancedLeft && balancedRight && abs(heightLeft - heightRight) <= 1;
    return std::max(heightLeft, heightRight) + 1;
}

bool isBalanced(Node* root)
{
    bool balanced;
    getHeight(root, balanced);
    return balanced;
}

int main(int argc, char* argv[])
{
    Node* tree1 = new Node{ 
        new Node{ 
            new Node{ 
                new Node{ nullptr, nullptr }, 
                new Node{ nullptr, nullptr } },
            new Node{ 
                new Node{ nullptr, nullptr },
                nullptr } },
        new Node{ 
            new Node{ 
                new Node{ nullptr, nullptr },
                new Node{ nullptr, nullptr } },
            new Node{ 
                new Node{ nullptr, nullptr },
                new Node{ 
                    new Node{ nullptr, nullptr },
                    nullptr } } } }; 

    bool b = isBalanced(tree1);

    Node* tree2 = new Node{ 
        new Node{ 
            new Node{ 
                new Node{ nullptr, nullptr }, 
                new Node{ nullptr, nullptr } },
            new Node{ 
                new Node{ nullptr, nullptr },
                nullptr } },
        new Node{ 
            new Node{ 
                new Node{ nullptr, nullptr },
                new Node{ nullptr, nullptr } },
            new Node{ 
                new Node{ nullptr, nullptr },
                new Node{ 
                    new Node{ new Node{ nullptr, nullptr }, nullptr },
                    nullptr } } } }; 

    b = isBalanced(tree2);

	return 0;
}

File: CrackingTheCodingInterview/4_trees_and_graphs/1_is_balanced.rb
class Node
  attr_accessor :child_left, :child_right

  def initialize(child_left, child_right)
    @child_left = child_left
    @child_right = child_right
  end
end

def get_balanced_and_height(root)
  return [true, 0] unless root

  left_balanced_and_height = get_balanced_and_height(root.child_left)   
  right_balanced_and_height = get_balanced_and_height(root.child_right)

  balanced = left_balanced_and_height[0] && right_balanced_and_height[0] && (left_balanced_and_height[1] - right_balanced_and_height[1]).abs <= 1
  height = [left_balanced_and_height[1], right_balanced_and_height[1]].max + 1

  [balanced, height]
end

def is_balanced(root)
  get_balanced_and_height(root)[0]
end


tree1 = Node.new( 
    Node.new( 
        Node.new( 
            Node.new( nil, nil ), 
            Node.new( nil, nil ) ),
        Node.new( 
            Node.new( nil, nil ),
            nil ) ),
    Node.new( 
        Node.new( 
            Node.new( nil, nil ),
            Node.new( nil, nil ) ),
        Node.new( 
            Node.new( nil, nil ),
            Node.new( 
                Node.new( nil, nil ),
                nil ) ) ) )

puts is_balanced(tree1)

tree2 = Node.new( 
    Node.new( 
        Node.new( 
            Node.new( nil, nil ), 
            Node.new( nil, nil ) ),
        Node.new( 
            Node.new( nil, nil ),
            nil ) ),
    Node.new( 
        Node.new( 
            Node.new( nil, nil ),
            Node.new( nil, nil ) ),
        Node.new( 
            Node.new( nil, nil ),
            Node.new( 
                Node.new( Node.new( nil, nil ), nil ),
                nil ) ) ) )

puts is_balanced(tree2)
